File: Seafood Guide/Seafood Guide/View Controller/FishTVC.m
//
//  FishTVC.m
//  Seafood Guide
//
//  Created by Jon Brown on 9/2/14.
//  Copyright (c) 2014 Jon Brown Designs. All rights reserved.
//

#import "FishTVC.h"
#import "CoreDataTableViewController.h"
#import "Seafood.h"
#import "AppDelegate.h"
#import "TVCDetailViewController.h"

@implementation FishTVC

@synthesize window = _window;
@synthesize fetchedResultsController = __fetchedResultsController;
@synthesize managedObjectContext = __managedObjectContext;


- (NSManagedObjectContext *)managedObjectContext
{
    if (!__managedObjectContext)
    {
        __managedObjectContext = [(AppDelegate *)[[UIApplication sharedApplication] delegate] managedObjectContext];
    }
    return __managedObjectContext;
    
}


- (void)setupFetchedResultsController

{
    
    // 1 - Decide what Entity you want
    NSString *entityName = @"Seafood"; // Put your entity name here
    //NSLog(@"Setting up a Fetched Results Controller for the Entity named %@", entityName);
    
    // 2 - Request that Entity
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:entityName];
    
    
//    NSString *theFrameThatWasTouchedwithTheUsersFinger = [[NSString alloc]init];
//    theFrameThatWasTouchedwithTheUsersFinger = note;

    
    // 3 - Filter it if you want

    
    NSUserDefaults * defaults = [NSUserDefaults standardUserDefaults];
    NSUInteger integerFromPrefs = [defaults integerForKey:@"justclicked"];
    predicateKey = [NSString stringWithFormat:@"%li",  (unsigned long)integerFromPrefs];
    
    if (integerFromPrefs == 0) {
        self.title = @"Mild Fish";
    } else if (integerFromPrefs == 1) {
        self.title = @"Thicker & Flavorful";
    } else if (integerFromPrefs == 2) {
        self.title = @"Steak Like";
    } else if (integerFromPrefs == 3) {
        self.title = @"Small Fish";
    } else if (integerFromPrefs == 4) {
        self.title = @"Shellfish";
    } else if (integerFromPrefs == 5) {
        self.title = @"Other";
    } else if (integerFromPrefs == 6) {
        self.title = @"All Fish";
    } else if (integerFromPrefs == 7) {
        self.title = @"Dirty Dozen";
    }
    
    //NSLog(@"Setting up predicate value %lu", (unsigned long)integerFromPrefs);
    
    if(predicateKey)
    {
        
        NSPredicate *pred = [NSPredicate predicateWithFormat:@"fishtype like[c] %@",predicateKey];
        [request setPredicate:pred];
        //NSLog(@"Setting up predicate value %@", predicateKey);
    }
    
    
    // 4 - Sort it if you want
    request.sortDescriptors = [NSArray arrayWithObject:[NSSortDescriptor sortDescriptorWithKey:@"name"
                                                                                     ascending:YES
                                                                                      selector:@selector(localizedCaseInsensitiveCompare:)]];
    
    // 5 - Fetch it
    self.fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                                                        managedObjectContext:self.managedObjectContext
                                                                          sectionNameKeyPath:@"uppercaseFirstLetterOfName"
                                                                                   cacheName:nil];
       
    
    
    [self.fetchedResultsController performFetch:nil];
    
}



- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [[self.fetchedResultsController sections] count];
}


- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
    return [sectionInfo name];
}

- (NSInteger)tableView:(UITableView *)table numberOfRowsInSection:(NSInteger)section {
    id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
    return [sectionInfo numberOfObjects];
}


- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self setupFetchedResultsController];
    
    
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    static NSString *CellIdentifier = @"FishCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }

    
    // Configure the cell...

    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    
    Seafood *seafood = [self.fetchedResultsController objectAtIndexPath:indexPath];
    
    cell.textLabel.text = seafood.name;

    
    return cell;
    
}

- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView
{
    return [[UILocalizedIndexedCollation currentCollation] sectionIndexTitles];
}

- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index {
    return [[UILocalizedIndexedCollation currentCollation] sectionForSectionIndexTitleAtIndex:index];
}

- (id)initWithFishSize:(NSString *)size
{
    self = [super initWithStyle:UITableViewStylePlain];
    
    return self;
}

- (id)initWithStyle:(UITableViewStyle)style
{
    return [self initWithFishSize:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navigationItem.backBarButtonItem.title = @"Back";
}

- (void)viewDidUnload
{
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

#pragma mark - Table view sends data to detail view
// Core data to detail view

-(void)tableView:(UITableView *)aTableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    TVCDetailViewController *detailViewController = [[TVCDetailViewController alloc]init];

    NSInteger sumSections = 0;
    for (int i = 0; i < indexPath.section; i++) {
        long rowsInSection = [self.tableView numberOfRowsInSection:i];
        sumSections += rowsInSection;
    }
    NSInteger currentRow = sumSections + indexPath.row;
    
    Seafood *allSeafood = [[self.fetchedResultsController fetchedObjects] objectAtIndex:currentRow];
    [detailViewController setItem:allSeafood];
    
    [self.navigationController pushViewController:detailViewController
                                         animated:YES];
    
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Back" style:UIBarButtonItemStylePlain target:nil action:nil];
}
-(void)viewDidAppear:(BOOL)animated
{
    [self.navigationItem setHidesBackButton:NO];
}

@end

File: Seafood Guide/Seafood Guide/View Controller/FishTVC.swift
//  Converted to Swift 5.1 by Swiftify v5.1.26565 - https://objectivec2swift.com/
//
//  FishTVC.swift
//  Seafood Guide
//
//  Created by Jon Brown on 9/2/14.
//  Copyright (c) 2014 Jon Brown Designs. All rights reserved.
//

import UIKit
import  CoreData

//#import "CoreDataTableViewController.h" // so we can fetch
//#import "Seafood.h"

class FishTVC: CoreDataTableViewController {
    var predicateKey = ""
    var uppercaseFirstLetterOfName = ""

    var window: UIWindow?

    lazy var managedObjectContext: NSManagedObjectContext? = {
        return (UIApplication.shared.delegate as? AppDelegate)?.managedObjectContext
    }()

    init(fishSize size: String?) {
        super.init(style: UITableView.Style.plain)
    }

    func prefersStatusBarHidden() -> Bool {
        return false
    }

    func setupFetchedResultsController() {

        let entityName = "Seafood" // Put your entity name here
        let request = NSFetchRequest<NSFetchRequestResult>(entityName: entityName)
        let defaults = UserDefaults.standard
        let integerFromPrefs = defaults.integer(forKey: "justclicked")
        predicateKey = String(format: "%li", UInt(integerFromPrefs))

        if integerFromPrefs == 0 {
            title = "Mild Fish"
        } else if integerFromPrefs == 1 {
            title = "Thicker & Flavorful"
        } else if integerFromPrefs == 2 {
            title = "Steak Like"
        } else if integerFromPrefs == 3 {
            title = "Small Fish"
        } else if integerFromPrefs == 4 {
            title = "Shellfish"
        } else if integerFromPrefs == 5 {
            title = "Other"
        } else if integerFromPrefs == 6 {
            title = "All Fish"
        } else if integerFromPrefs == 7 {
            title = "Dirty Dozen"
        }

        if predicateKey != "" {

            let pred = NSPredicate(format: "fishtype like[c] %@", predicateKey)
            request.predicate = pred
        }

        request.sortDescriptors = [NSSortDescriptor.init(key: "name", ascending: true)]
        if let managedObjectContext = managedObjectContext {
            fetchedResultsController = NSFetchedResultsController(fetchRequest: request, managedObjectContext: managedObjectContext, sectionNameKeyPath: "uppercaseFirstLetterOfName", cacheName: nil)
        }



        do {
            try fetchedResultsController?.performFetch()
        } catch {
        }

    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return fetchedResultsController?.sections?.count ?? 0
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        weak var sectionInfo = fetchedResultsController?.sections?[section]
        return sectionInfo?.name
    }

    override func tableView(_ table: UITableView, numberOfRowsInSection section: Int) -> Int {
        weak var sectionInfo = fetchedResultsController?.sections?[section]
        return sectionInfo?.numberOfObjects ?? 0
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        setupFetchedResultsController()


    }

    static let tableViewCellIdentifier = "FishCell"

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        var cell = tableView.dequeueReusableCell(withIdentifier: FishTVC.tableViewCellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: FishTVC.tableViewCellIdentifier)
        }

        cell?.accessoryType = .disclosureIndicator

        let seafood = fetchedResultsController?.object(at: indexPath) as? Seafood

        cell?.textLabel?.text = seafood?.name


        return cell!

    }

    override func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return UILocalizedIndexedCollation.current().sectionIndexTitles
    }

    override func tableView(_ tableView: UITableView, sectionForSectionIndexTitle title: String, at index: Int) -> Int {
        return UILocalizedIndexedCollation.current().section(forSectionIndexTitle: index)
    }

    convenience override init(style: UITableView.Style) {
        self.init(fishSize: nil)
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.backBarButtonItem?.title = "Back"
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

// MARK: - Table view sends data to detail view
    // Core data to detail view
    override func tableView(_ aTableView: UITableView, didSelectRowAt indexPath: IndexPath) {

        let detailViewController = TVCDetailViewController.init(style: .plain)

        var sumSections = 0
        for i in 0..<indexPath.section {
            let rowsInSection = tableView.numberOfRows(inSection: i)
            sumSections += rowsInSection
        }
        let currentRow = sumSections + indexPath.row

        let allSeafood = fetchedResultsController?.fetchedObjects?[currentRow] as? Seafood
        detailViewController.item = allSeafood

        navigationController?.pushViewController(detailViewController, animated: true)

        navigationItem.backBarButtonItem = UIBarButtonItem(title: "Back", style: .plain, target: nil, action: nil)
    }

    override func viewDidAppear(_ animated: Bool) {
        navigationItem.hidesBackButton = false
    }
}

//#import "CoreDataTableViewController.h"
